utils/buffer/README.md

```diff
@@ -1,105 +1,151 @@
+- **Slice**: A contiguous block of memory, either preallocated or dynamically allocated.
 
-## Implementation
-The crate is `[no_std]` and lock-free, so, to synchronize the state of the pre-allocated memory
-(taken or free) between different contexts an `AtomicU8` is used.
-Each bit of the `u8` represent a memory slot, if the bit is 0 the memory slot is free if is
-1 is taken. Whenever `BufferPool` creates a `Slice` a bit is set to 1 and whenever a `Slice` is
-dropped a bit is set to 0.
+## Usage
 
-## Use case
-`BufferPool` has been developed to be used in proxies with thousand of connection, each connection
-must parse a particular data format via a `Decoder` each decoder use 1 or 2 `Buffer` for each received
-message. With `BufferPool` each connection can be instantiated with its own `BufferPool` and reuse
-the space freed by old messages for new ones.
+To include this crate in your project, run:
 
-## Unsafe
-There are 5 unsafes:
-buffer_pool/mod.rs 550
-slice.rs 8
-slice.rs 27
+```bash
+cargo add buffer_sv2
+```
+
+This crate can be built with the following feature flags:
+
+- `debug`: Provides additional tracking for debugging memory management issues.
+- `fuzz`: Enables support for fuzz testing.
+- `with_serde`: builds [`binary_sv2`](https://crates.io/crates/binary_sv2) and
+  [`buffer_sv2`](https://crates.io/crates/buffer_sv2) crates with `serde`-based encoding and
+  decoding. Note that this feature flag is only used for the Message Generator, and deprecated
+  for any other kind of usage. It will likely be fully deprecated in the future.
+
+### Unsafe Code
+There are four unsafe code blocks instances:
+
+- `buffer_pool/mod.rs`: `fn get_writable_(&mut self, len: usize, shared_state: u8, without_check: bool) -> &mut [u8] { .. }` in the `impl<T: Buffer> BufferPool<T>`
+- `slice.rs`:
+  - `unsafe impl Send for Slice {}`
+  - `fn as_mut(&mut self) -> &mut [u8] { .. }` in the `impl AsMut<[u8]> for Slice`
+  - `fn as_ref(&mut self) -> &mut [u8] { .. }` in the `impl AsMut<[u8]> for Slice`
+
+### Examples
+
+This crate provides three examples demonstrating how the memory is managed:
+
+1. **[Basic Usage Example](https://github.com/stratum-mining/stratum/blob/main/protocols/v2/codec-sv2/examples/basic_buffer_pool.rs)**:
+   Creates a buffer pool, writes to it, and retrieves the data from it.
 
-## Write
-Waiting for `Write` in `core` a compatible trait is used so that it can be replaced.
+2. **[Buffer Pool Exhaustion Example](https://github.com/stratum-mining/stratum/blob/main/protocols/v2/codec-sv2/examples/buffer_pool_exhaustion.rs)**:
+   Demonstrates how data is added to a buffer pool and dynamically allocates directly to the heap
+   once the buffer pool's capacity has been exhausted.
 
-## Buffer
-The `Buffer` trait has been written to work with `codec_sv2::Decoder`.
+3. **[Variable Sized Messages Example](https://github.com/stratum-mining/stratum/blob/main/protocols/v2/codec-sv2/examples/variable_sized_messages.rs)**:
```

Comment:
these links are all pointing as if the examples lived under `protocols/v2/codec-sv2`